websockets/websockets_test.go

```diff
@@ -879,3 +880,241 @@ func TestCookiesDefaultJar(t *testing.T) {
 	assert.Equal(t, map[string]string{"someheader": "defaultjar"}, collected)
 	mu.Unlock()
 }
+
+func TestCompression(t *testing.T) {
+	t.Parallel()
+
+	t.Run("session", func(t *testing.T) {
+		t.Parallel()
+		const text string = `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas sed pharetra sapien. Nunc laoreet molestie ante ac gravida. Etiam interdum dui viverra posuere egestas. Pellentesque at dolor tristique, mattis turpis eget, commodo purus. Nunc orci aliquam.`
+
+		ts := newTestState(t)
+		sr := ts.tb.Replacer.Replace
+		ts.tb.Mux.HandleFunc("/ws-compression", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			upgrader := websocket.Upgrader{
+				EnableCompression: true,
+				ReadBufferSize:    1024,
+				WriteBufferSize:   1024,
+			}
+
+			conn, e := upgrader.Upgrade(w, req, w.Header())
+			if e != nil {
+				t.Fatalf("/ws-compression cannot upgrade request: %v", e)
+				return
+			}
+
+			// send a message and exit
+			if e = conn.WriteMessage(websocket.TextMessage, []byte(text)); e != nil {
+				t.Logf("error while sending message in /ws-compression: %v", e)
+				return
+			}
+
+			e = conn.Close()
+			if e != nil {
+				t.Logf("error while closing connection in /ws-compression: %v", e)
+				return
+			}
+		}))
+
+		err := ts.ev.Start(func() error {
+			_, runErr := ts.rt.RunString(sr(`
+		// if client supports compression, it has to send the header
+		// 'Sec-Websocket-Extensions:permessage-deflate; server_no_context_takeover; client_no_context_takeover' to server.
+		// if compression is negotiated successfully, server will reply with header
+		// 'Sec-Websocket-Extensions:permessage-deflate; server_no_context_takeover; client_no_context_takeover'
```

Comment:
we don't probably, actually I've just left as a reminder (but probably a bad reminder) that we still need to figure out a way how this can be responded (probably to the WebSocket.extensions) :thinking: